### ISSUE TITLE
re-timing keys and drawing at the same

### ISSUE THREAD

**Issue by adamearle** [opentoonz/opentoonz#3429]
Workflow issue with spacing drawings and keys at the same time in the Timeline and function editor. 

At the moment any animation either done tradigital or cutout animation in retiming is very tedious and time-consuming. Through ongoing use over the years working with all kinds of artists the below list is are much-needed refinements of OpenToonz

Retiming animation should not be as painstaking as it is in OpenToonsz. We believe in the following request that workflow will speed up dramatically. This means that there will be a tighter faster cleaner way to navigate the timeline/ Xsheet & Function (Curve), editor. No more late nights on retiming animations because of OpenToonz outdated use of current modern code and computers. Animators' life quality will become significantly better from a few refinements.

We don't believe the Timeline/ Xsheet and Function (Curve) editor needs to be rewritten simply refined.

Function Editor: (Curve Editor)
> Retime everything across the timeline and Moving and the curve editor
> Select multiple keyframes at once by clicking dragging
> Remove select curve capability
> Remove all Key Spacing options and have just one this can be broken up in to 
    1. Drawing 2. Keys Icons that can be switched on separately or both at the same time
> Spacing out keyframes and drawings "Add space" & "Remove Space" across multiple columns
> Shift Keys backwards and forwards - Need to be changed to "shift keys Forward"/ "backward"

> An option to Reveal in the function editor from the

> Remove double click from the Timeline/ Xsheet to open the function editor and make it Ctrl + double click
> Copy-pasting keys + Hot Keys 
> Solo Keys + Hot Keys 
> Mute keys + Hot Keys 
> Isolating the selection + Hot Keys 
> Invert selection on columns + Hot Keys 

> Preference: Animation / (the option to have broken handles or not in the curve editor) 
> Make the keys at the top larger in the Curve Editor
> The repeat icon could also be seen in the curve editor
> Quick Switching icon so windows can be quickly switched between timelines/ Xsheet Function Editor

> Curve editor needs a normalized view of the curves
> Zoom in and out around on canvas during playback

> Curve editor appearance needs an update so that 

Timeline:
> Just changing and adding these 3 columns means that  retiming could all be done in the timeline
>        1.Add Mask column  
>        2. Add Peg column  
>        3. Add FX column

> Play every "x"frames
> Draw and animate only on odd/ evens

> Sync Timelines this would offer the ability to see multiple timelines open at a time with only  columns that can be seen each timeline  so isolation of retiming can that retiming can happen in a more modern clear and speeder 
> Multiple Timeline open with different selection (Maybe a pin option)
> Remove resequence this should just be an automatic this procedure computers can handle it now.

> Key placement on columns can be altered so that moving keys and drawing on the timeline can happen more easily.

> Show and hide all
> Show and hide Selected only 
> Show and hide Columns only
> Show and hide Pegs only 
> Show and hide Camera only
> Show and hide Table only
> Show and hide Audio only
> Show and hide Pegs only (Column needs to be added)
> Show and hide Mask only (Column needs to be added)
> Show and hide FX only (Column needs to be added)
> Show and hide Notes only
>        Jump to notes

> Add & Delete Markers
>        Marker Naming
>        Marker Commenting

> Middle click draw to switch drawing frames out from the level editor could open up the level editor if it is not open or pop up the level editor window

**Comment by adamearle** [comment]
@shun-iwasawa I am looking outside help. 

What are your thoughts on this? 

Is there anything in that you think you can tackle?

Cheers Ad

**Comment by shun-iwasawa** [comment]
Thank you @adamearle for suggestions.
Some items seem to be relatively-easily doable and some are not. Let me check...

**Comment by DarrenTAnims** [comment]
Just to clarify 3 of my biggest issues with re-timing on the timeline/xsheet (and I know others have the same issues here)

1. I'd like to be able to drag a key without double clicking. At the minute you have to click once to select a key, then click and drag to move it. You really should be able to just click and drag it.

2. Doing this double clicking to move a key often results in the function editor popping up. This is so annoying! I never want to popup the function editor. But I'm sure some do, so could this not just move to using a modifier key to pop it up, Alt+Click?

3. When I re-time the drawings, either by dragging with the frame drag handles or by inserting frames, I want the keys to move too. I can't think of a single time where the keys are so detached from the drawings that I don't want them to move with the drawings. These are the keys on xsheet, from the animate tool, and the keys in the function editor, from the fx.

If these could be considered alongside the above, that'd be great.
Cheers,
Darren

**Comment by artisteacher** [comment]
> When I re-time the drawings, either by dragging with the frame drag handles or by inserting frames, I want the keys to move too. I can't think of a single time where the keys are so detached from the drawings that I don't want them to move with the drawings. These are the keys on xsheet, from the animate tool, and the keys in the function editor, from the fx.

I very much agree with this one! While we can use ctrl/cmd to select the drawings & keys at the same time, in most cases trying to move the selection ends up deselecting the keys. It's super annoying. There probably are cases where the key does need to stay in place, but that is likely the minority. 

I'd like to see the normal behavior be that keys stick to the drawings when moving. Then a modifier would be used when moving drawings to get the current behavior (keys locked to the frame number).

> Play every "x"frames

Isn't that short play? You can set the number of frames in Preferences under Preview. Then you just need to set a shortcut. Maybe there should be a toggle in the play controls to toggle it on & off though...

**Comment by adamearle** [comment]
@artisteacher if this is what you mean it needs to exposed more in the UI. 

![image](https://user-images.githubusercontent.com/16330340/90711284-a7ddb580-e2e3-11ea-996d-3c840d200082.png)


Also, some projects demand that everything gets baked down to 1s 2s 3s sometimes it depends on the studio. I know my self and other artist work on 1s 2s 3s and more to get the timing out fast and then come back in to and worry about if it on 1s or 2s

**Comment by adamearle** [comment]
Yeah that needs to be in the timeline really doesn't it. @artisteacher
